### ISSUE TITLE
SSP Completeness Checks: 7 External Systems and Services Not Having FedRAMP Authorization

### ISSUE THREAD

**Issue by brian-ruf** [GSA/fedramp-automation#808]
### This is a ...

fix - something needs to be different

### This relates to ...

- [ ] the **FedRAMP OSCAL Registry**
- [ ] the **FedRAMP OSCAL baselines**
- [ ] the **Guide to OSCAL-based FedRAMP Content**
- [X] the **Guide to OSCAL-based FedRAMP System Security Plans (SSP)**
- [ ] the **Guide to OSCAL-based FedRAMP Security Assessment Plans (SAP)**
- [ ] the **Guide to OSCAL-based FedRAMP Security Assessment Results (SAR)**
- [ ] the **Guide to OSCAL-based FedRAMP Plan of Action and Milestones (POA&M)**
- [X] the **FedRAMP SSP OSCAL Template** (JSON or XML Format)
- [ ] the **FedRAMP SAP OSCAL Template** (JSON or XML Format)
- [ ] the **FedRAMP SAR OSCAL Template** (JSON or XML Format)
- [ ] the **FedRAMP POA&M OSCAL Template** (JSON or XML Format)
- [X] the **FedRAMP OSCAL Validations**

### User Story

As a consumer of FedRAMP automated completeness checks I want the following OSCAL-based SSP items to be automatically verified for completeness by metaschema constraints:
- Check for presence
- If present, check for required fields

### Goals

SSP Completeness checks are defined, tested and documented

### Acceptance Criteria

- [ ] All FedRAMP Documents Related to OSCAL Adoption (https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated.
- [ ] A Pull Request (PR) is submitted that fully addresses the goals of this User Story. This issue is referenced in the PR.
- [ ] all constraints associated with the review task have been converted/created
- [ ] [automate.fedramp.gov](http://automate.fedramp.gov/) content has been updated accordingly
- [ ] the metaschema help prop has an appropriate link to the constraint
- [ ] the template has an content that models the desired OSCAL presentation
- [ ] the constraint runs against the example template
- [ ] known-bad content has been created
- [ ] the constraint appropriately flags the known-bad content as invalid

**Comment by brian-ruf** [comment]
There are four scenarios that require tracking as part of _External Systems and Services Not Having FedRAMP Authorization_:

**Scenario 1** FedRAMP-Authorized System connecting to an external system that may or may not be FedRAMP-authorized.
- Create one //component[@type='system'] for each external system
- Create one //component[@type='interconnection'] for each connection between this system and the external system
- Link the interconnection component to the external system using the component UUID of the external system in: `//component[@type='service']/prop[@name='connected-to-system-uuid'][@ns='http://fedramp.gov/ns/oscal']/@value`
- Link the interconnection component to this system using the component UUID of the this systemin : `//component[@type='service']/prop[@name='connected-to-system-uuid'][@ns='http://fedramp.gov/ns/oscal']/@value`
- Provide Interconnection Security Agreement details
- Provide remote and local connection details
- Provide nature of agreement details
- Provide information type details
- Provide risk details

**Scenario 2** FedRAMP-Authorized System using a service from an underlying leveraged system, where the underlying system **is** FedRAMP-authorized; however, the service **is not** included in the underlying system's authorization. (The service is not included in the underlying's system's FedRAMP Marketplace details.)
- Create one //component[@type='service'] for each service used
- Unlike the use of services falling under the leveraged system's authorization, the above component must **NOT** have prop[@name='leveraged-authorization-uuid']. 
  - This would incorrectly indciate the service is covered under the authorization. It is not. 
- Link the service component to the leveraged system using the component UUID of the leveraged system in: `//component[@type='service']/prop[@name='offered-by-system-uuid'][@ns='http://fedramp.gov/ns/oscal']/@value`
- Provide information type details
- Provide risk details


**Scenario 3** FedRAMP-Authorized System using an API or service from an external system:
- Create one //component[@type='system'] for each external system
- Create one //component[@type='service'] for each service or API used
- Link the API or service component to the leveraged system using the component UUID of the leveraged system in: `//component[@type='service']/prop[@name='offered-by-system-uuid'][@ns='http://fedramp.gov/ns/oscal']/@value`
- Provide nature of agreement details
- Provide remote and local connection details
- Provide information type details
- Provide risk details

**Scenario 4** FedRAMP-Authorized System using an CLI to manage remote or underlying system:
- Create one //component[@type='system'] for each system where the CLI can effect change 
- Create one //component[@type='software'] for each CLI used
- Link the software component to the leveraged system using the component UUID of the leveraged system in: `//component[@type='service']/prop[@name='offered-by-system-uuid'][@ns='http://fedramp.gov/ns/oscal']/@value`
- Provide nature of agreement details
- Provide remote and local connection details
- Provide information type details
- Provide risk details

**Comment by brian-ruf** [comment]
# STILL WORKING HERE

TARGET: 
- `//system-implementation/component[((@type=('system', 'service')) and not(./prop[@name='leveraged-authorization-uuid']) and ./prop[@name='implementation-point' and @value='external']) or (@type='interconnection') or (@type='software' and ./prop[@name='asset-type' and @value='cli'])]`


| Data | Location | Documentation Notes | Example Updated |
|  :--- |  :--- | :--- | :---: | 
| #  |  n/a / Sequential Numbering | n/a | n/a | n/a | 
| System/Service/API/CLI Name [1]               | `./title` (required field) | Good | Y | 
| Service Processor Name  [1]                   | `./responsible-role[@name='provider']/party-uuid`  | | Y | 
| Connection Details (Direction) [1 or 2]       | `./prop[@name='direction'    and @ns='http://fedramp.gov/ns/oscal']/@value`   |  | Y | 
| Connection Details (Local IPv4 Address) [0+]  | `./prop[@name='ipv4-address' and @class='local' and @ns='http://fedramp.gov/ns/oscal']/@value`   |  | Y | 
| Connection Details (Local IPv6 Address) [0+]  | `./prop[@name='ipv6-address' and @class='local' and @ns='http://fedramp.gov/ns/oscal']/@value`   |  | Y | 
| Connection Details (Local Port) [0+]          | `./protocol[@name='local']/port-range`   |  | Y | 
| Connection Details (Remote IPv4 Address) [0+] | `./prop[@name='ipv4-address' and @class='remote' and @ns='http://fedramp.gov/ns/oscal']/@value`   |  | Y | 
| Connection Details (Remote IP6 Address) [0+]  | `./prop[@name='ipv6-address' and @class='remote' and @ns='http://fedramp.gov/ns/oscal']/@value`   |  | Y | 
| Connection Details (Remote Port) [0+]         | `./protocol[@name='remote']/port-range`     |  | Y | 
| Connection Details (Security )                | `./prop[@name='direction' and @ns='http://fedramp.gov/ns/oscal']/@value`   |  | Y | 
| Nature of Agreement [1]                       | `./prop[@name='nature-of-agreement' and @ns='http://fedramp.gov/ns/oscal']/@value` |  | Y | 
| Still Supported? [1]                          | `./prop[@name='still-supported' and @ns='http://fedramp.gov/ns/oscal']`  |  | Y | 
| Data Types [1+]                               | `./prop[@name='data-type' and @ns='http://fedramp.gov/ns/oscal']/@value` |  | Y | 
| Data Categorization [1+]                      | `./prop[@name='data-categorization' and @ns='http://fedramp.gov/ns/oscal']/@value` |  | Y | 
| Authorized Users [0+]                         | `./responsible-role[@role-id='authorized-users']/party-uuid`  |   | Y | 
| User Authentication Method [0 or 1]           | `./prop[@name='user-authentication' and @ns='http://fedramp.gov/ns/oscal']`  |   |  Y | 
| Other Compliance Programs [0+]                | `./prop[@name='compliance-program' and @ns='http://fedramp.gov/ns/oscal']/@value`  |  | Y | 
| Description [1] OSCAL Required                | `./description` |  | Y | 
| Hosting Environment [1]                       | `./prop[@name='hosting-environment' and @ns='http://fedramp.gov/ns/oscal']`   |  |Y  | 
| Risk [0+]                                     | `./prop[@name='risk' and @ns='http://fedramp.gov/ns/oscal']`   |  |Y  | 
| Impact [0+]                                   | `./prop[@name='impact' and @ns='http://fedramp.gov/ns/oscal']`   |  |Y  | 
| Mitigation [0+]                               | `./prop[@name='mitigation' and @ns='http://fedramp.gov/ns/oscal']`   |  |Y  |

**Comment by brian-ruf** [comment]
# STILL WORKING HERE

## Constraints Needed:
- If `./prop[@name='direction' and @ns='http://fedramp.gov/ns/oscal' and @value='incoming']`
  - If direction is incoming there must be at least one local ipv4 or ipv6 address.
    - `count(./prop[@name=('ipv4-address', 'ipv6-address') and @class='local' and @ns='http://fedramp.gov/ns/oscal']/@value) >=1`
  - If direction is incoming there must be at least one local ports entry.
    - `count(./protocol[@name='local']/port-range/@start)`

- If `./prop[@name='direction' and @ns='http://fedramp.gov/ns/oscal' and @value='outgoing']`
  - If direction is outgoing there must be at least one remote ipv4 or ipv6 address.
    - `count(./prop[@name=('ipv4-address', 'ipv6-address') and @class='remote' and @ns='http://fedramp.gov/ns/oscal']/@value) >=1`
  - If direction is outgoing there must be at least one remote ports entry.
    - `count(./protocol[@name='remote']/port-range/@start)`

- `count(./responsible-role[@name='provider']/party-uuid) = 1`: Exactly one provider party
- `./responsible-role[@name='provider']/party-uuid ` exists in `//part` index

**Comment by brian-ruf** [comment]
# NIST OSCAL Issues Blocking This Issue
- [ ] usnistgov/OSCAL#2023
- [ ] usnistgov/OSCAL#2044

**Comment by brian-ruf** [comment]
@Telos-sa you actually caught a mistake. It was supposed to be "provided-by", not "offered-by". I've corrected. "provided-by" is part of the core OSCAL syntax.

Yes the goals here are as follows:
- use core OSCAL syntax as much as possible.
- keep the data constructs the same as much as possible between 6.1 (LAs) and 7.1 (external and unauthorized)
- Keep the data normalized as much as possible, which includes referring to the to information from the leveraged system on behalf of the service, rather than duplicating that information into the service.

It's still a work in progress, but I'm working on one example for each of the above scenarios here:
[https://github.com/brian-ruf/fedramp-automation/blob/example-ssp/src/content/rev5/examples/ssp/xml/FedRAMP-SSP-Example.OSCAL.xml](https://github.com/brian-ruf/fedramp-automation/blob/example-ssp/src/content/rev5/examples/ssp/xml/FedRAMP-SSP-Example.OSCAL.xml)

**Comment by brian-ruf** [comment]
Changed `user-authentication` property/extension to `authentication-method`.

**Comment by brian-ruf** [comment]
Modifying the "user-authentication" property/extension to an "authentication-method" property/extension as the true intention of this information is to document the nature of any authentication method, whether it be external users or external systems/automation accounts.

**Comment by aj-stein-gsa** [comment]
> Changed `user-authentication` property/extension to `authentication-method`.

@brian-ruf, just a heads up because I can tell this may be the same thing as the leveraged authorization prop will likely be the same thing, that prop elsewhere is already wrapping on a review. If we wanted to change that, we may want to notify Gabe as he is working on https://github.com/GSA/fedramp-automation/issues/891.

**Comment by brian-ruf** [comment]
@aj-stein-gsa thank you. Yes, I actually needed to make notes/revisions to PR #921, as well as issues #807, #891, #893, #897, and created issue #924.

The good news is this will check boxes in both 807 and 808. While I did have to tweak the language associated with the allowed values, it turned out that language was missing from the PR anyway. So we can get the updated language in when that gets corrected.

**Comment by brian-ruf** [comment]
After consultation with the review team as captured in #935, we have eliminated "risk", "mitigation", and "impact" property/extensions in favor of a "poam-item" link. This is based on the assertion that every entry in Table 7.1 is required to have a corresponding POA&M entry. 

The primary benefits of referring to the POA&M entry can include:
- elimination of redundancy
- simplicity of data modeling and tool development
- the POA&M model offers far more comprehensive treatment of risk and mitigation vs property/extensions

**Comment by brian-ruf** [comment]
We need allowed values for the "connection-security" prop/extension. Created issue #961 to address this.

**Comment by brian-ruf** [comment]
For "service" or "software" components with an "implementation-point" of "external" it is clear that communication crosses the boundary.

For "service" and "software" components  I am adding a "communicates-externally" prop/extension that is required when the "implementation-point" is "internal". This is how tools will know to impose Table 7-1 information requirements on these internal components. (information types, authentication, POA&M item, etc.)

**Comment by brian-ruf** [comment]
In the wake of recent revisions, and in an effort to reduce future rework, I am moving to the following approach for things like properties and links that should appear in some components, but not others based on complex conditions:
- Have one constraint that enforces the presence of the field (or fields) based on those complex conditions.
- Have very simply allowed value constraints that simply enforce allowed values in any component where the property or link exists. 

In other words, allowed value constraints should always trust that cardinality or complex circumstances are enforced by other constraints and are not the "responsibility" of the allowed value constraint. This should also simplify error messages.

**Comment by Telos-sa** [comment]
> For "service" or "software" components with an "implementation-point" of "external" it is clear that communication crosses the boundary.
> 
> For "service" and "software" components I am adding a "communicates-externally" prop/extension that is required when the "implementation-point" is "internal". This is how tools will know to impose Table 7-1 information requirements on these internal components. (information types, authentication, POA&M item, etc.)

For this new prop "communicates-externally", what are the expected values of this new prop?  should it be a "yes"/"no"?

**Comment by brian-ruf** [comment]
@Telos-sa, yes. That is correct. There will be an allowed values constraint that defines/enforces, "yes" and "no" as the two allowed values for "communicates-externally"
